### ISSUE TITLE
[feature-request] node aliases

### ISSUE THREAD

**Issue by pchdev** [ossia/score#346]
up for debate, and following the idea developed in jamoma if I'm not mistaken: allow a node to have an name/address alias, in order to make some unchangeable osc addresses clearer.

i.e.: in reaper, I have a parameter I want to modulate at the address "/track/1/fx/1/fxparameter/2/value" and I'd like to assign it a clearer name.

**Comment by bltzr** [comment]
If I understand that well, does that mean that you permanently want to address this with e.g. /some/address in i-score (while the actual OSC address Reaper expects is /track/1/fx/1/fxparameter/2/value) ?
If so, maybe a more general idea would be to be able to have a kind of OSC mapping, where we could substitute a tree in i-score to the OSC tree the remote device understands
there could also be some pattern-matching magic, in order to "tidy up" some namespaces (e.g. Dlight, if my memory is good) and make them more relevant to the way things are managed in i-score
I think this was discussed in the first OSSIA workshop in Lyon, and that we took notes about this
maybe we should add this to the discussions to address in Montbéliard... @theod @reno- ?

**Comment by pchdev** [comment]
@bltzr yes, exactly!

**Comment by theod** [comment]
> maybe we should add this to the discussions to address in Montbéliard... 

yes ! a very first feature would be to have a new node type called "alias" that allow to point to another node in existing devices.
so you don't allow a node to have one alias but you allow to create as many aliases you need for one node. this sounds a more powerful feature to me.

**Comment by jcelerier** [comment]
I've added it, now gotta use it in i-score. 
https://github.com/OSSIA/API/blob/master/OSSIA/ossia/network/generic/alias_node.hpp

**Comment by bltzr** [comment]
maybe a way to use it would be by adding a alias field just below the name (when editing a node)
then the node shows up in the device tree with the alias (instead of its "built-in" address
if it has children, they are nested under this aliased address (?)
opinions, @OSSIA/users ?

**Comment by pchdev** [comment]
hmm, ok for me

/track/1/fx/1 would have an alias for example : /reaper/filter

now, say you want to give /track/1/fx/1/fxparameter/2/value an alias (/reaper/filter/cutoff)
you click add child on /reaper/filter
maybe type fxparameter/2/value for the absolute node adress
and then just “cutoff“ for the alias 

is that suitable?

**Comment by bltzr** [comment]
what do you mean by "cutting off the alias" ?

**Comment by bltzr** [comment]
oh sorry, I misunderstood :-D

**Comment by bltzr** [comment]
well why not... I guess that's mainly a question for @jcelerier if that's doable ?

**Comment by bltzr** [comment]
IMHO, what would be best, would be that subnodes can inherit from their parent's (or ancestor's) alias, but that they can have their own particular aliases too.

In the given example:
the user gives `/track/1/fx/1 ` the alias `/reaper/filter`
its subnode `/track/1/fx/1/fxparameter/2/value` is automatically given the alias `/reaper/filter/fxparameter/2/value`
which they can replace (by editing this node's alias) to `/reaper/filter/cutoff`
(we can help the user on this by automatically selecting the specific part in the editor, i.e. `fxparameter/2/value`)
or even, if they want it so, to `/my/amazing/cutoff`, thus having this node's alias existing outside of the parent's alias sub-hierarchy 
would that make sense to you ?

**Comment by pchdev** [comment]
yes, even better!

**Comment by jcelerier** [comment]
> the user gives /track/1/fx/1 the alias /reaper/filter its subnode /track/1/fx/1/fxparameter/2/value is automatically given the alias /reaper/filter/fxparameter/2/value

isn't there a risk of a ton of duplicated nodes ?

**Comment by bltzr** [comment]
why/how ?

**Comment by jcelerier** [comment]
> its subnode /track/1/fx/1/fxparameter/2/value is automatically given the alias /reaper/filter/fxparameter/2/value

The address " /reaper/filter/fxparameter/2/value" would be visible from the device explorer right ?

**Comment by bltzr** [comment]
nope, it would be replaced by its alias IMO

**Comment by bltzr** [comment]
we can consider having two display modes for the tree: absolute or aliases

**Comment by jcelerier** [comment]
> nope, it would be replaced by its alias IMO

ok, that's a big difference :p this should go in the "device explorer work package" then I think.

**Comment by bltzr** [comment]
sure, done - though, maybe that can be done in a second batch, once the main refactor is done, right? or should it be considered from the start of the refactor (if so I'll move it to the 1.0 milestone)

**Comment by jcelerier** [comment]
I think that it should be considered from the start - it may be a big change

**Comment by jcelerier** [comment]
(but I *really* doubt it will be done by then)

**Comment by jcelerier** [comment]
would the current "Mapper" device be enough to cover this FR ? (reminder of how it works : https://github.com/OSSIA/score-user-library/blob/master/Devices/Mapper/Mapper.qml)
